torchtune/datasets/_packed.py

```diff
@@ -0,0 +1,274 @@
+
+        # Only show progress bar on rank 0
+        _, rank = get_world_size_and_rank()
+        if rank == 0:
+            pbar = tqdm(total=len(self.ds), desc="Packing dataset", dynamic_ncols=True)
+
+        for batch in self.ds:
+            tokens, labels = batch["tokens"], batch["labels"]
+            # If the dataset outputs samples that are larger than the specified
+            # max_seq_len and we're unable to split it, user needs to modify
+            # one of the two parameters
+            seq_len = len(tokens)
+            if seq_len > self.max_seq_len and not self.split_across_pack:
+                raise ValueError(
+                    f"Dataset sample is too long ({len(tokens)} > {self.max_seq_len}). "
+                    "Please set `split_across_pack=True` or increase `max_seq_len`."
+                )
+
+            # Create integer mask and position ids for current sample and extend
+            # current pack
+            current_sample = {
+                "tokens": tokens,
+                "labels": labels,
+                # Mask is simply a causal mask within this sample length
+                "mask": [torch.tril(torch.ones(seq_len, seq_len, dtype=torch.bool))],
+                "input_pos": list(range(seq_len)),
+            }
+            current_pack = {k: v + current_sample[k] for k, v in current_pack.items()}
+
+            # If the current pack is long enough, add it to self.packs and retain
+            # any truncated samples for next pack, if splitting samples
+            if len(current_pack["tokens"]) > self.max_seq_len:
+                current_pack = self._add_pack(
+                    current_pack=current_pack,
+                    boundary=self.max_seq_len
+                    if self.split_across_pack
+                    else previous_sample_boundary,
+                )
+
+            if rank == 0:
+                pbar.update()
+            previous_sample_boundary = len(current_pack["tokens"])
+            if self.max_packs is not None and len(self.packs) >= self.max_packs:
+                break
+
+        # Add the last pack with remaining samples that did not fit in previous
+        if len(current_pack["tokens"]) > 0 and (
+            self.max_packs is None or len(self.packs) < self.max_packs
+        ):
+            current_pack = self._add_pack(
+                current_pack=current_pack, boundary=len(current_pack["tokens"])
+            )
+            assert len(current_pack["tokens"]) == 0
+
+    def _add_pack(
+        self, current_pack: Dict[str, List[int]], boundary: int
+    ) -> Dict[str, List[int]]:
+        """
+        Pad and add the current pack to self.packs and return what's remaining.
+        """
+        # So far we've kept a list of causal masks, one for each sample in the pack.
+        # Now we need to combine them into a single mask for the entire pack.
+        packing_mask = torch.block_diag(*current_pack["mask"])
+        pack = {
```

Comment:
Not sure if it'll be simpler to manage this `pack` as a dataclass but might be worth considering.

Reply:
the code will still be ugly, won't it? since I have to assign to each field. and I want to return them as dicts in the batch